File: precond.hs
type PVar = String
data Term = Nil | Bar String | Hat String
data Pred = EQ Term Term | NE Term Term
type PathExp = [Pred]
type SymbStore = [(PVar, Term)]
type PureFml = (SymbStore, PathExp)
data SpatPred = Mapsto Term Term | Ls Term Term
type SpatFml = [SpatPred]
type SymbHeap = (PureFml, SpatFml)

--type PVar = String
type FunName = String

data Exp = NULL | Var String
-- data SafeComm = Assign PVar Exp | Malloc PVar
-- data UnSafeComm = DerefAssign PVar Exp | Modify Exp Exp | Free Exp
data BExp = Eeq Exp Exp | Ene Exp Exp
data Comm = Assign PVar Exp | Malloc PVar -- Safe Comm
          | Derefer PVar Exp | Modify Exp Exp | Free Exp -- Unsafe Comm
          | Call FunName [Exp] | If BExp Prog Prog
type Prog = [Comm]

data FunDef = Fun FunName [PVar] Prog

apply :: SymbStore -> Exp -> Term
apply s NULL = Nil
apply s (Var x) = maybe undefined $ lookup x s

pre :: Comm -> (SymbStore, PathExp, SpatFml, SpatFml) -> [(SymbStore, PathExp, SpatFml, SpatFml)]
pre (Assign x e) (s, d, heap, foot) = return (((x, e'):s), d, heap, foot) where
  e' = apply s e
pre (Malloc x) (s, d, heap, foot) = return ((x, Hat a):s, (Hat a `NE` Nil):d, (Hat a `Mapsto` Hat a'):heap, foot)
pre (Derefer x e) (s, d, heap, foot) | allocated(d, heap, e) = undefined
pre (Derefer x e) (s, d, heap, foot) = undefined
pre (Modify e e') (s, d, heap, foot) | allocated(d, heap, e) = undefined
pre (Modify e e') (s, d, heap, foot) = undefined
pre (Free e) (s, d, heap, foot) | allocated(d, heap, e) = undefined
pre (Free e) (s, d, heap, foot) = undefined
pre (Call f es) (s, d, heap, foot) = undefined
pre (If b p1 p2) (s, d, heap, foot) = undefined

preProg :: Prog -> (SymbStore, PathExp, SpatFml, SpatFml) -> [(SymbStore, PathExp, SpatFml, SpatFml)]
preProg [] a = a
preProg (c:p) (s, d, h, f) = do
  (s', d', h', f') <- pre c (s, d, h, f)
  preProg p (s', d', h', f')

main = print undefined

File: precond.ml
open Types

(* apply : symbStore -> exp -> term *)
let rec apply s = function
  | NULL -> Nil
  | Var x -> List.assoc x s

(* applyB : symbStore -> bExp -> predicate *)
let applyB s = function
  | Eeq (e1, e2) -> EQ (apply s e1, apply s e2)
  | Ene (e1, e2) -> NE (apply s e1, apply s e2)

(* neg : predicate -> predicate *)
let neg = function
  | EQ (t1, t2) -> NE (t1, t2)
  | NE (t1, t2) -> EQ (t1, t2)

(* mapsto_of : spatFml -> (term * term) list *)
let mapsto_of = List.map (function
                    | Mapsto (t1, t2) -> (t1, t2)
                    | Ls (t1, t2) -> (t1, Hat "end_of_ls")) (* TODO: tilde var *)

(* dom : spatFml -> term list *)
let dom h = List.map fst (mapsto_of h)

(* allocated : pathExp -> spatFml -> term -> bool *)
let allocated d h e = List.mem e (dom h) (* TODO: Use pathexp *)

(* rearrange : pathExp -> spatFml -> term -> term *)
let rearrange d h e = List.assoc e (mapsto_of h)

(* gensym : unit -> string *)
let counter = ref 0
let gensym () =
  counter := !counter + 1;
  "a" ^ string_of_int !counter

(* expand_ls : pathExp -> spatFml -> term -> spatFml *)
let expand_ls d h e = h         (* TODO: impl this *)

(* modify_heap : spatFml -> term -> term -> spatFml*)
let modify_heap h e e' = List.map (function
                             | Mapsto (e1, e2) when e1 = e -> Mapsto (e1, e')
                             | Mapsto (e1, e2) -> Mapsto (e1, e2)
                             | Ls (e1, e2) -> Ls (e1, e2)) h

let string_of_result sdhf_list =
  String.concat "\n" (List.map (fun (s, d, h, f) ->
                          "{" ^ string_of_pureFml (s, d) ^ ", "
                          ^ string_of_spatFml h ^ ", "
                          ^ string_of_spatFml f ^ "}") sdhf_list)

let debug_sdhf sdhf = print_endline (string_of_result sdhf)

let add_footprint (s,d,h,f) e xbar =
  (s, d, Mapsto (e, xbar) :: h, Mapsto (e, xbar) :: f)

(* pre : comm -> symbStore * pathExp * spatFml * spatFml -> (symbStore * pathExp * spatFml * spatFml) list *)
(* preProg : prog -> SH -> SH list *)
let rec pre c (s, d, h, f) = match c with
  | Assign (x, e) ->
     let e' = apply s e in [((x, e') :: s, d, h, f)]
  | Malloc x ->
     let ahat = Hat (gensym ()) and
         ahat' = Hat (gensym ()) in
     [(x, ahat)::s, NE (ahat, Nil) :: d, Mapsto (ahat, ahat') :: h, f]
  | Derefer (x, e) when allocated d h (apply s e) ->
     let e' = apply s e in
     let e'' = rearrange d h e' in
     let h' = expand_ls d h e' in
     [(x, e'')::s, d, h', f]
  | Derefer (x, e) ->
     let es = apply s e in
     let xbar = Bar (gensym ()) in
     let (s, d, h, f) = add_footprint (s, d, h, f) es xbar in
     [(x, xbar)::s, List.map (fun e -> NE (es, e)) (dom h) @ d, h, f]
  | Modify (e, e') when allocated d h (apply s e) ->
     let es = apply s e and
         e's = apply s e' in
     let h' = modify_heap h es e's in
     [s, d, h', f]
  | Modify (e, e') ->
     let es = apply s e and
         e's = apply s e' in
     let e'' = rearrange d h es in
     let xbar = Bar (gensym ()) in
     [add_footprint (s, d, h, f) es xbar]
  | Free e when allocated d h (apply s e) -> failwith "free"
  | Free e -> failwith "free"
  | Call (f, es) -> failwith "call"
  | If (b, p1, p2) -> let b' = applyB s b in
                      preProg p1 (s, b' :: d, h, f) @ preProg p2 (s, (neg b') :: d, h , f)
and preProg p (s, d, h, f) = match p with
  | [] -> [(s, d, h, f)]
  | c :: p -> let rs = pre c (s, d, h, f) in
              print_string (string_of_comm c); debug_sdhf rs;
              List.concat (List.map (fun sdhf -> preProg p sdhf) rs)

(* disp-list (x) { if x == NULL then return else y = *x; disp_list(y); free(x) *)
let dispose_list_body = [
    Malloc "x";
    If (Eeq ((Var "x"), NULL),
      [],
      [Derefer ("y", Var "x"); Call ("disp_list", [Var "y"]); Free (Var "x")])
  ]
(* swap(x, y) = { local t; t = *x; s = *y; *x = s; *y = t; } *)
let swap_body = [
    Derefer ("t", Var "x");
    Derefer ("s", Var "y");
    Modify (Var "x", Var "s");
    Modify (Var "y", Var "t")
  ]

let initSymbStore = ["x", Bar "x"; "y", Bar "y"; "t", Hat "t"; "s", Hat "s"]

let main () =
  print_string (string_of_prog swap_body);
  print_endline "-----";
  debug_sdhf [initSymbStore, [], [], []];
  let res = preProg swap_body (initSymbStore, [], [], []) in
  res
let _ = main ()
